ls1intum/Artemis
# `Programming exercises`: Add possibility to configure docker run arguments on bamboo via spring properties

### Checklist
#### General


(https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
#### Server


### Motivation and Context


Programming Exercises may negatively influence the Build Infrastructure if not prevented explicitly. Especially C programming exercises can crash Build Agents with memory leaks or fork bombs. Since the exercises are executed within a docker container, we can limit the containers resources explicitly and kill the container before the agent crashes. 


### Description

> This change only impacts bamboo based deployments at the moment 

In this PR we add a new configuration option for [docker run arguments](https://docs.docker.com/engine/reference/run/): `artemis.continuous-integration.build.default-docker-flags`. An Administrator may provide any docker run flag as a key-value pair to this config. Artemis will explicitly add those flags to all newly created Bamboo Build plans. 

We chose a default limit of 2 CPUs, 2GB Ram and 1000 pids. These values should be sufficient for most of the exercises. If they are not, instructors have to adapt the base and template build plan in bamboo with more resources (and respective bamboo "requirements" to only use more powerful agents). 

Acknowledgements: 
- @sleiss: Helping with Spring magic 
- @RobertJndw and @kliwniloc: Help debugging the build infrastructure 

### Steps for Testing

Prerequisites:
- 1 Instructor

__Steps in Artemis__

1. Log in to Artemis
2. Navigate to Course Administration
3. Select a Course 
4. Create a new Programming Exercise 
5. Open the base and solution build plans in bamboo (links can be found in the lower half of the exercise overview page)

__Steps in Bamboo__

6. Click on "Actions" and then on "Configure plan" in the top right corner of the build plan 
7. Select the "Default Job" 
8. Change to the "Docker" Tab 
9. Inspect the "Extra arguments". They should look like this: 

```
--cpus=2
--memory=2g
--pids-limit=1000
--memory-swap=2g
```
10. Check that both build plans build successfully 


### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test
